linkerd/app/inbound/src/detect.rs

```diff
@@ -63,22 +73,34 @@ impl<N> Inbound<N> {
         const TLS_PORT_SKIPPED: tls::ConditionalServerTls =
             tls::ConditionalServerTls::None(tls::NoServerTls::PortSkipped);
 
-        self.map_stack(|cfg, rt, tls| {
+        self.map_stack(|cfg, rt, detect| {
             let detect_timeout = cfg.proxy.detect_protocol_timeout;
-            tls.check_new_service::<Tls, tls::server::Io<I>>()
-                .push_request_filter(
-                    |(tls, t): (tls::ConditionalServerTls, T)| -> Result<Tls, DeniedUnauthorized> {
+            detect
+                .push_switch(
+                    // Ensure that the connection is authorized before proceeding with protocol
+                    // detection.
+                    |(tls, t): (tls::ConditionalServerTls, T)| -> Result<_, Error> {
                         let policy: AllowPolicy = t.param();
                         let permit = policy.check_authorized(tls)?;
-                        Ok(Tls::from_params(&t, permit))
+
+                        // If the port is configured to support application TLS, it may have also
+                        // been wrapped in mesh identity. In any case, we don't actually validate
+                        // whether app TLS was employed, but we use this as a signal that we should
+                        // not perform additional protocol detection.
```

Comment:
do we want to eventually have this configuration validate that the app is using TLS?